graviti/utility/user.py

```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class OrderedDict(MutableMapping[str, _V], UserMapping[str, _V]):
+    """OrderedDict is a dict of ordered elements, supports searching the element by its index."""
+
+    __marker = object()
+
+    def __init__(
+        self, values: Union[Iterable[Tuple[str, _V]], Mapping[str, _V], None] = None
+    ) -> None:
+        self._keys: List[str] = []
+        self._data: Dict[str, _V] = {}
+
+        if not values:
+            return
+
+        iterable = values.items() if isinstance(values, Mapping) else values
+        for key, value in iterable:
+            self._keys.append(key)
+            self._data[key] = value
+
+    def __getitem__(self, key: Union[int, str]) -> _V:
+        if isinstance(key, int):
+            key = self._keys.__getitem__(key)
+
+        return self._data.__getitem__(key)
+
+    def __setitem__(self, key: str, value: _V) -> None:
+        if not self._data.__contains__(key):
+            self._keys.append(key)
+
+        self._data.__setitem__(key, value)
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        if isinstance(key, int):
+            self._keys.__delitem__(key)
+            key = self._keys.__getitem__(key)
+        else:
+            self._keys.remove(key)
+
+        self._data.__delitem__(key)
+
+    def __contains__(self, key: Any) -> bool:
+        return self._data.__contains__(key)
+
+    def __iter__(self) -> Iterator[str]:
+        return self._keys.__iter__()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self._data.__eq__(other._data) and self._keys.__eq__(other._keys)
+
+    def keys(self) -> Tuple[str, ...]:  # type: ignore[override]
+        """Get all element keys in order.
+
+        Returns:
+            A tuple containing all elements keys.
+
+        """
+        return tuple(self._keys)
+
+    def items(self) -> Tuple[Tuple[str, _V], ...]:  # type: ignore[override]
```

Comment:
shall we return an `Iterator`?

Reply:
If `items()` returns `Iterator`, should `keys()` and `values()` also return `Iterator`?
In the current situation, there wouldn't be too many items in it.

Reply:
obviously Iterator is better than tuple